Index generator

eccentric-j opened the pull request:
- Created scripts/index-imgs.js using primitives from the tsxify-svgs.js in #4 
- It works intelligently so that after initial creation, it only generates indexes for folders where a file has changed

## Todos
- [x] Cleanup fromBinder implementation -- Feels like it could be simpler
- [x] Update the first-run logic in both sync and watch mode, right now they read all img files, get the directory, put them in a set, and then operate on each directory in the set. Ideally, it should just stream a list of directories. Don't expect this to be too big, should be able to configure glob & chokidar options
- [x] ~~Don't generate indexes if no img files exist in that directory~~ Wait, already does that 😛 
- [x] Update package.json to use index-imgs script

jmezzacappa commented:
This is cool :)
Aside: Chokidar supports globs, so we may not need the `glob` dependency. (Unless I overlooked something... I barely scanned through the code, but the index files are awesome!)

eccentric-j commented:
> This is cool :) Aside: Chokidar supports globs, so we may not need the `glob` dependency. (Unless I overlooked something... I barely scanned through the code, but the index files are awesome!)

Yeah I am using chokidar's glob feature but as far as I could tell, when trying to run it once without `--watch` chokidar stays open so the script doesn't exit, even with persistent: false. I can experiment more with that though and see if I can get rid of glob

eccentric-j commented:
Did a quick test, persistent: false works as expected so should be able to get rid of glob

eccentric-j commented:
Got all those todo items done.

- Initial run only emits directories instead of individual imgs
- In watch, emits directories of each changed image
- Fully lazy, never creates any arrays of files or directory names in memory
- Moved the stream primitives into another dir just to make the actual indexing code not so scary looking
- Updated some UI in healthcare and government page, wanted to avoid merge conflicts

Also can confirm the latest svgr configs fixed svgs so was able to update the government page to show clients without any loaders.

jmezzacappa commented on scripts/lib/stream.js:
```diff
@@ -0,0 +1,207 @@
+const { Readable, Transform, PassThrough } = require('stream')
```
I'm surprised this thing required using streams directly 😳 Are we transforming large images? Or maybe some excessive IO (e.g. inadvertently indexing `node_modules` recursively or something) happening somewhere?

eccentric-j replied:
It's not processing large images or node_modules or anything, it's not even reading the images' contents. I just think they're useful for scripts like these. 

There's going to be about 20 pages each with about 5 or more images. The svgr script already takes about a second to run and is only going to get slower where as this script will keep it quite fast and efficient no matter how many files we throw at it.

Personally, I also like having some kind of structure that conveys the whole story from top to bottom:

```js
streamImgFiles('./app/route-assets/**/*.{svg,png,jpg}')
  .pipe(stream.map(directory => ({
    directory,
    index: path.join(directory, 'index.ts'),
    contents: readdirStream(directory)
      .pipe(
        stream.filter(file =>
          isImgFile(['svg', 'png', 'jpg', 'tsx'], file)
        ),
      )
      .pipe(stream.map(formatImgExport)),
  })))
  .pipe(createIndexWriter())
  .on('finish', () => {
    console.log(`Generated indexes`)
  })
  .on('error', console.error)
```

I also like that no matter how many times it uses map, filter, once, etc... it only fires once per file unlike arrays which will iterate n times, which I think would become more noticeable once we start getting more pages in. 

Could be done with a library like https://gulpjs.com/ but comes with a bigger footprint in terms of packages, and I'm open to another stream library, my favorite for these kinds of scripts used to be https://www.npmjs.com/package/highland as it had good built-in support for native node streams, but it has not been maintained.

eccentric-j replied:
Plus it was the weekend so wanted to have some fun with it 😁

eccentric-j replied:
Sorry, one last thought: I didn't make the stream functions for this project. Made them for tsxify-svgs.js and reused them here since they are versatile building blocks we can compose together for these kinds of problems.

For example, I think I'd like to write another script that uses chokidar for images, then  runs them through svgr API directly. That way instead of it processing every file every time there's a single change, it will generate them once on start (which is good if we end up making changes to an svg directly like removing the clip-path), then only update the files that are changed, added, or removed after that. Only a small amount of code needs to be written to accomplish that as I can reuse those primitives and model it roughly the same way.

jmezzacappa replied:
Chokidar actually has what we need to do all that. Check out how much simpler this could be:

```js
// @ts-check

const fs = require('fs/promises')
const path = require('path')
const chokidar = require('chokidar')

const GLOB = './app/route-assets/**/*.{svg,png,jpg,ts,tsx}'

const flags = {
	watch: process.argv.includes('--watch') || process.argv.includes('-w'),
}

/**
 * Transforms a string like cloud_hosted into IconCloudHosted
 *
 * @param {string} basename - A basename string without extension "always_on"
 * @returns {string}
 */
function pascalCase(basename) {
	return basename
		.split(/[-_\s]+/)
		.map(word => {
			const char = word[0]
			const chars = word.slice(1)
			return `${char.toUpperCase()}${chars}`
		})
		.join('')
}

/**
 * @param {string} dirname
 * @param {string[]} filenames
 */
async function writeIndex(dirname, filenames) {
	const indexFileContents = filenames.reduce(
		(acc, filename) => {
			const extname = path.extname(filename)
			const basename = path.basename(filename, extname)

			let exportLine = ''
			switch (extname) {
				case '.png':
				case '.jpg':
				case '.svg':
					exportLine = `export { default as img${
						pascalCase(basename)
					} } from './${filename}'\n`
					break

				case '.ts':
				case '.tsx':
					exportLine =
						`export { default as ${basename} } from './${basename}'\n`
			}

			return acc.concat(exportLine)
		},
		'',
	)
	if (!indexFileContents)
		return

	const indexFilePath = path.join(dirname, 'index.ts')
	await fs.writeFile(indexFilePath, indexFileContents, { encoding: 'utf-8' })

	console.log('Created index file', indexFilePath)
}

const reIgnore = /\/?index.tsx?$/

const watcher = chokidar.watch(GLOB, {
	persistent: flags.watch,
	ignoreInitial: true,
	ignored: reIgnore,
})

watcher.on('ready', () => {
	const promises = Object
		.entries(watcher.getWatched())
		.map(([dirname, filenames]) => writeIndex(dirname, filenames))

	Promise.all(promises).then(() => console.log('Generated indexes'))
})

if (flags.watch) {
	watcher
		.on('add', listener)
		.on('change', listener)
		.on('unlink', listener)

	/** @param {string} filePath */
	function listener(filePath) {
		const dirname = path.dirname(filePath)
		fs.readdir(dirname)
			.then(filenames => filenames.filter(name => !reIgnore.test(name)))
			.then(filenames => writeIndex(dirname, filenames))
	}
}
```

AFAICT, it checks all the boxes you mentioned (including only running once per file), but with far less code and abstractions, no direct use of streams, type safe, and no additional dependencies (e.g. Gulp, Highlander, etc.). It's also very fast.

It took less than an hour to write, and I could also throw the SVGR thing in there if it helps.

I understand that streams can be fun, but using them where they're not necessary can be exhausting for others ([example](https://twitter.com/BenLesh/status/1438926352896434177)).

eccentric-j replied:
> AFAICT, it checks all the boxes you mentioned (including only running once per file, asynchronous, etc.), but with far less code and abstractions, no direct use of streams, type safe, and no additional dependencies (e.g. Gulp, Highlander, etc.). It's also very fast.

I don't think it's good that it has to load every dir's filenames. It's also far less efficient as it loads the entire tree into memory and create an io promise for every file at once.

> Also, native JS streams are designed more for data streams (IO stuff). Making them work as event streams is a bit much, IMO.

I'm using them for IO stuff. It starts with chokidar, then finishes using fs.writeFileStream. It's not an uncommon connection given https://github.com/floatdrop/gulp-watch, https://www.npmjs.com/package/@proscom/gulp-svgr, and https://www.npmjs.com/package/gulp-svgo.

> It took less than an hour to write, and I could also throw the SVGR thing in there if it helps.

Spent not that much on these scripts either. The svgr one I pushed last night also took less than an hour to write with those building blocks in place.

> I understand that streams can be fun, but using them where they're not necessary can be exhausting for others ([example](https://twitter.com/BenLesh/status/1438926352896434177)).

I've read that article before. I'm not using them to make things complicated or over-engineer, but because I think they're the right tool for the job to make a high quality toolset for these types of problems. The goal is to avoid memory spikes as I've run into issues with containers being reaped processing too many items in memory at once during build time. It's a bit of abstraction, but it builds on familiar concepts like array methods.

In Elixir, it comes with stream primitives https://hexdocs.pm/elixir/1.12/Stream.html, in modern Java it also comes with chainable stream methods now since Java 8 https://www.tutorialspoint.com/java8/java8_streams.htm, and in Clojure laziness is built into the collection processing functions that chunk arrays out once they process 32+ items. This is the exact type of problem they are well suited for and I don't see the value in treating them like some resource to avoid when they are demonstrably working and faster than their synchronous counterparts like in the case of svgr.

jmezzacappa replied:
> I don't think it's good that it has to load every dir's filenames.

I just checked, and the largest array of filenames is 30 strings.
If we someday have a single dir with too many files to list in memory, then it would make more sense to optimize.

> It's also far less efficient as it loads the entire tree into memory

I'm not sure that's accurate. The streaming approach doesn't make Chokidar watch the file tree any differently, AFAICT.
Maybe I'll check the memory consumption of both to see if it's an issue.

>  and create an io promise for every file at once.

What problem does this create for us that NodeJS doesn't already handle?

FWIW, the fixture record/playback app does this far more heavily, and it hasn't caused any issues at all.
They're promises. We're basically just telling NodeJS "Hey, here's a list of async stuff to do. Let me know when you're done". It seems to handle it pretty well.

If this _does_ become an actual issue someday, some sort of worker/pooler thing could still be implemented with far less code.

I suppose an "incremental build" implementation could make that more efficient, but that's an optimization that both solutions could benefit from at some point anyway.

> It starts with chokidar, then finishes using fs.writeFileStream.

Same, pretty much... But with way less code as `fs.writeFile` takes care of the IO part. It's just a one-liner:
`fs.writeFile(indexFilePath, indexFileContents, { encoding: 'utf-8' })`

> I've read that article before.

It's actually just a Tweet by Ben Lesh, the RxJS core team lead.

> I'm not using them to make things complicated or over-engineer, but because I think they're the right tool for the job to make a high quality toolset for these types of problems.

This seems like the kind of stuff that would force us into using some heavy, restrictive Ruby framework. The goal was to build a website, but now we're slipping down a rabbit hole where we're building some platform for making toolsets.

Custom build tooling has been such a massive time-sink. It's like we're iterating on hammer designs just to hang a picture. At this point, I almost seems worthwhile to just forego all of the SVG processing stuff, and go back to maintaining/editing the SVGs by hand again. Simpler times.

The irony here is that the purpose of this script is to write _less_ code.

eccentric-j replied:
> It's actually just a Tweet by Ben Lesh, the RxJS core team lead.

Yes I know who he is, he was the primary dev behind RxJS, worked at Netflix before joining Microsoft to work on RxJS full time. He's covered that topic in articles, his talks, and that tweet.

> This seems like the kind of stuff that would force us into using some heavy, restrictive Ruby framework. The goal was to build a website, but now we're slipping down a rabbit hole where we're building some platform for making toolsets.

That was written during the weekend to smooth out the development process and now we're spending too much time on it. Now we don't get to use any of it and we're wasting time on this because you're blocking the PR. Why not just use them because they work then you can simplify later?

eccentric-j replied:
Summarizing our meeting

- Let's merge them for now so we can use these tools
- Can switch to Joey's version once we get the green light on this prototype

Going forward, will refrain from reaching for these as they're unfamiliar and may be overkill for these problems 😅
